controllers/rollout/canary.go

```diff
@@ -0,0 +1,269 @@
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and upgrade workload new versions", canaryStatus.CurrentStepIndex, steps)
+	r.newStatus.Message = cond.Message
+	// promote workload next batch release
+	if batchState.Paused || batchState.CurrentBatch < canaryStatus.CurrentStepIndex {
+		klog.Infof("rollout(%s/%s) will promote batch from(%d) -> to(%d)", r.rollout.Namespace, r.rollout.Name, batchState.CurrentBatch, canaryStatus.CurrentStepIndex)
+		return false, r.batchControl.PromoteBatch(canaryStatus.CurrentStepIndex)
+	}
+
+	// check whether batchRelease is ready
+	if batchState.State == batchrelease.BatchInRollingState {
+		klog.Infof("rollout(%s/%s) workload(%s) batch(%d) availableReplicas(%d) state(%s), and wait a moment",
+			r.rollout.Namespace, r.rollout.Name, r.workload.Name, canaryStatus.CurrentStepIndex, batchState.UpdatedReadyReplicas, batchState.State)
+		return false, nil
+	}
+	klog.Infof("rollout(%s/%s) workload(%s) batch(%d) availableReplicas(%d) state(%s), and continue",
+		r.rollout.Namespace, r.rollout.Name, r.workload.Name, canaryStatus.CurrentStepIndex, batchState.UpdatedReadyReplicas, batchState.State)
+	return true, nil
+}
+
+func (r *rolloutContext) doCanaryMetricsAnalysis() (bool, error) {
+	// todo
+	return true, nil
+}
+
+func (r *rolloutContext) doCanaryPaused() (bool, error) {
+	// No step set, need manual confirmation
+	if len(r.rollout.Spec.Strategy.Canary.Steps) == 0 {
+		klog.Infof("rollout(%s/%s) don't contains steps, and need manual confirmation", r.rollout.Namespace, r.rollout.Name)
+		return false, nil
+	}
+	canaryStatus := r.newStatus.CanaryStatus
+	currentStep := r.rollout.Spec.Strategy.Canary.Steps[canaryStatus.CurrentStepIndex-1]
+	steps := len(r.rollout.Spec.Strategy.Canary.Steps)
+	cond := util.GetRolloutCondition(*r.newStatus, appsv1alpha1.RolloutConditionProgressing)
+	// need manual confirmation
+	if currentStep.Pause.Duration == nil {
+		klog.Infof("rollout(%s/%s) don't set pause duration, and need manual confirmation", r.rollout.Namespace, r.rollout.Name)
+		cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and you need manually confirm(kube-cli approve) to enter the next step", canaryStatus.CurrentStepIndex, steps)
+		r.newStatus.Message = cond.Message
+		return false, nil
+	}
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and wait duration(%d seconds) to enter the next step", canaryStatus.CurrentStepIndex, steps, *currentStep.Pause.Duration)
+	r.newStatus.Message = cond.Message
+	// wait duration time, then go to next step
+	duration := time.Second * time.Duration(*currentStep.Pause.Duration)
+	expectedTime := canaryStatus.LastUpdateTime.Add(duration)
+	if expectedTime.Before(time.Now()) {
+		klog.Infof("rollout(%s/%s) canary step(%d) paused duration(%d seconds), and go to the next step",
+			r.rollout.Namespace, r.rollout.Name, canaryStatus.CurrentStepIndex, *currentStep.Pause.Duration)
+		return true, nil
+	} else {
```

Comment:
else block can be removed here, I see this pattern in many places, is this on purpose?

Reply:
Yes, because the current state is not as expected, it takes time to re-reconcile